feat(executioncontext): warn on nested js handle

bgschiller opened the pull request:
ExecutionContext.evaluateHandle accepts arguments that are either
serializable, or JSHandles. A potential confusion is that it *does not*
accept arguments that *contain* JSHandles.

This patch adds a log message warning when it encounters that situation.

Fixes #3562

bgschiller committed:
feat(ExecutionContext): warn on nested js handle

ExecutionContext.evaluateHandle accepts arguments that are either
serializable, or JSHandles. A potential confusion is that it *does not*
accept arguments that *contain* JSHandles.

This patch adds a log message warning when it encounters that situation.

Fixes #3562

SimonSchick commented:
I don't know if that's a good thing to do, if anything this should re-throw as another error, emitting warnings on incorrect usage that *should* result in errors doesn't make much sense.

I think it would be sufficient to alter the error message eg. `err.message += ' Are you passing a nested JSHandle?'`

aslushnikov commented:
I like @SimonSchick suggestion - that's the right amount of code/effort for the problem we're having.

bgschiller commented:
Sounds good, I'll make that change and push an update

bgschiller committed:
feat(ExecutionContext): warn about nested JSHandle

remove the code checking if args contain a nested JSHandle, and
instead simply suggest the possibility.

SimonSchick commented on lib/ExecutionContext.js:
```diff
@@ -111,15 +111,21 @@ class ExecutionContext {
+        userGesture: true
+      });
+    } catch (err) {
+      err.message += ' Are you passing a nested JSHandle?';
```
You *do* want to check if it's a TypeError still.

bgschiller replied:
Ah, right. I'll add that back.

SimonSchick reviewed (approved):
This is better but I'm still unsure if this sort of validation is the right way to do at all.
There are probably hundreds possible ways to pass bad data into puppeteer and make it give obscure errors, that's (to some degree) due to the nature of JS.

If you want actual type safety, you want to go with something like TypeScript.

bgschiller commented:
@SimonSchick, I'd absolutely prefer to lean on the type system for this. I actually *am* using Typescript, but I don't get an error in this case. Looking at `puppeteer/index.d.ts`, I see the following definition:

```ts
evaluateHandle(
  fn: EvaluateFn,
  ...args: any[]
): Promise<JSHandle>;
```

Even though args is documented as being `Serializable | JSHandle`, the type definitions don't include that information.

I'm pretty new to Typescript, but I'm not aware of a way to specify that arguments be serializable. Maybe you can point me in the right direction? Is there a built-in type I'm missing, or would it involve a definition something like this?

```ts
type Serializable = number | boolean | string | { [key: string]: Serializable } | Serializable[];
```

bgschiller committed:
only warn on TypeError

aslushnikov commented on lib/ExecutionContext.js:
```diff
@@ -111,15 +111,22 @@ class ExecutionContext {
+        userGesture: true
+      });
+    } catch (err) {
+      if (err instanceof TypeError && err.message === 'Converting circular structure to JSON')
```
this error happens due to `convertArgument` call. Let's scope it down to this:

```js
try {
  args = args.map(convertArgument.bind(this));
} catch (err) {
  if (err instanceof TypeError && err.message === 'Converting circular structure to JSON')
    err.message += ' Are you passing a nested JSHandle?';
  throw err;
}
```

bgschiller replied:
Unfortunately, the error actually happens after convertArgument, in `CDPSession.send`. It looks like convertArgument prepares the args for serialization, but the call to JSON.stringify is later on.

aslushnikov commented on lib/ExecutionContext.js:
```diff
@@ -162,6 +169,7 @@ class ExecutionContext {
         throw new Error('Execution context was destroyed, most likely because of a navigation.');
       throw error;
     }
+
```
nit: stray line

aslushnikov reviewed (commented):
Thanks; can you please also add a test?

bgschiller committed:
feat(ExecutionContext): warn on nested js handle

Test that the error thrown includes the altered message

bgschiller committed:
remove trailing comma for node6

bgschiller commented:
As far as I can tell, test failures are either unrelated (`1) Browser Page Workers Page.workers (worker.spec.js:8:5)`) or due to linting in files I haven't edited. Do I need to rebase/merge a more recent commit?

Also, I learned how to write a "Serializable" type, so I'm intending to submit a PR to DefinitelyTyped as well.

aslushnikov committed:
nit: naming

aslushnikov reviewed (approved):
Awesome, thanks!
